core/arch/arm/plat-imx/pm/psci-cpuidle-imx7.S

```diff
@@ -0,0 +1,773 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright 2018 NXP
+ */
+
+#include <kernel/unwind.h>
+#include <asm.S>
+#include <arm.h>
+#include <arm32_macros.S>
+#include <generated/imx_pm_asm_defines.h>
+#include <imx-regs.h>
+#include <platform_config.h>
+#include <kernel/cache_helpers.h>
+#include <kernel/tz_ssvce_def.h>
+#include <kernel/tz_proc_def.h>
+
+#define MX7_SRC_GPR1	0x74
+#define MX7_SRC_GPR2	0x78
+#define MX7_SRC_GPR3	0x7c
+#define MX7_SRC_GPR4	0x80
+#define MX7_GPC_IMR1	0x30
+#define MX7_GPC_IMR2	0x34
+#define MX7_GPC_IMR3	0x38
+#define MX7_GPC_IMR4	0x3c
+#define DDRC_STAT	0x4
+#define DDRC_PWRCTL	0x30
+#define DDRC_DBG1	0x304
+#define DDRC_DBGCAM	0x308
+#define DDRC_PSTAT	0x3fc
+#define DDRC_PCTRL_0	0x490
+
+	.section .text.psci.cpuidle
+	.align 3
+/*
+ * imx_pen_lock
+ *
+ * The reference link of Peterson's algorithm:
+ * http://en.wikipedia.org/wiki/Peterson's_algorithm
+ *
+ * val1 = r1 = !turn (inverted from Peterson's algorithm)
+ * on cpu 0:
+ * r2 = flag[0] (in flag0)
+ * r3 = flag[1] (in flag1)
+ * on cpu1:
+ * r2 = flag[1] (in flag1)
+ * r3 = flag[0] (in flag0)
+ *
+ */
+	.macro	imx_pen_lock
+
+	mov	r8, r0
+	read_mpidr r5
+	and	r5, r5, #3
+	add	r6, r8, #PM_INFO_MX7_VAL_OFF
+	cmp	r5, #0
+	addeq	r7, r8, #PM_INFO_MX7_FLAG0_OFF
+	addeq	r8, r8, #PM_INFO_MX7_FLAG1_OFF
+	addne	r7, r8, #PM_INFO_MX7_FLAG1_OFF
+	addne	r8, r8, #PM_INFO_MX7_FLAG0_OFF
+
+	mov	r9, #1
+	str	r9, [r7]
+	dsb
+	str	r5, [r6]
+1:
+	dsb
+	ldr	r9, [r8]
+	cmp	r9, #1
+	ldreq	r9, [r6]
+	cmpeq	r9, r5
+	beq	1b
+
+	.endm
+
+	.macro	imx_pen_unlock
+
+	dsb
+	read_mpidr r6
+	and	r6, r6, #3
+	cmp	r6, #0
+	addeq	r7, r0, #PM_INFO_MX7_FLAG0_OFF
+	addne	r7, r0, #PM_INFO_MX7_FLAG1_OFF
+	mov	r9, #0
+	str	r9, [r7]
+
+	.endm
+
+	.macro  disable_l1_dcache
+
+	push	{r0 - r12, lr}
+	mov	r0, #DCACHE_OP_CLEAN_INV
+	ldr	r1, =dcache_op_all
+	mov	lr, pc
+	bx	r1
```

Comment:
Why not `blx r1` instead?

Reply:
I could not recall clearly. But seems blx is better here. I'll fix in next version for all such case.